Add OSC133 prompt marking support

dcolascione opened the pull request:
This change teaches terminal emulators to understand the logical flow of REPL-ish Ink-based-app sessions by (a)busing OSC133 prompt and command output markers: we make the entire non-static UI a giant OSC133 "prompt" and each block of static output an OSC133 "command output".

This way, in terminal emulators that support the
feature (e.g. iTerm2), these markers allow users to click specific conversation turns to highlight them, use keyboard shortcuts to scroll between them, and so on, just as users can in the shell.

Terminal emulators that don't support OSC133 just ignore our marker sequences.

See https://iterm2.com/documentation-escape-codes.html

Test: start the program under iTerm2 and click the outputs.

dcolascione committed:
Add OSC133 prompt marking support

This change teaches terminal emulators to understand the logical flow
of REPL-ish Ink-based-app sessions by (a)busing OSC133 prompt and
command output markers: we make the entire non-static UI a giant
OSC133 "prompt" and each block of static output an OSC133 "command
output".

This way, in terminal emulators that support the
feature (e.g. iTerm2), these markers allow users to click specific
conversation turns to highlight them, use keyboard shortcuts to scroll
between them, and so on, just as users can in the shell.

Terminal emulators that don't support OSC133 just ignore our
marker sequences.

See https://iterm2.com/documentation-escape-codes.html

Test: start the program under iTerm2 and click the outputs.

wolffiex commented on src/components/Text.tsx:
```diff
@@ -53,6 +54,11 @@ export type Props = {
 	readonly children?: ReactNode;
+
+        /**
+          * Whether this text is semantically a prompt.
```
can you expand this comment with some details about osc133? minor suggestion, but I prefer like 
 `isOsc133Prompt` in JS

dcolascione replied:
I'd generally prefer it that way too, but I'm trying to be consistent with the rest of this data structure.

wolffiex commented on src/ink.tsx:
```diff
@@ -17,6 +17,11 @@ import Yoga from 'yoga-wasm-web/auto';
 
 const noop = () => {};
 
+const oscPromptStartRefreshLine = '\x1b]133;A\x07';
+const oscPromptEnd = '\x1b]133;B\x07';
+const oscCommandStart = '\x1b]133;C\x07';
+const oscCommandEnd = '\x1b]133;D\x07';
```
This could use a comment that points to OSC 133 documentation

dcolascione replied:
Done

wolffiex reviewed (commented):
The code here looks fine. I'm a little concerned that we are introducing ANSI escape sequences after points where we maybe thought it was sanitized, but I'll guess we'll find out if this breaks CI or something. I'll leave it @bcherny to accept—he's our resident ink expert.

Commit:
Address PR feedback

 - don't emit control codes in CI
 - add more doc comments

bcherny commented:
Thanks Dan! Testing this out:

1. I'm seeing this arrow in iTerm2 -- can we hide it? What can we do here?
<img width="46" alt="Screenshot 2025-03-06 at 10 21 03 AM" src="https://github.com/user-attachments/assets/7ba8c240-5b71-49b7-a4f1-94c703e8640b" />

2. Also seeing weird highlighting when I click on the terminal. Any way to avoid this?:
<img width="636" alt="Screenshot 2025-03-06 at 10 22 22 AM" src="https://github.com/user-attachments/assets/2033ff29-6061-4625-942c-d9287bc6faa6" />

3. Colors seem to have stopped working. eg. orange in the spinner is not rendering, and red in error messages:
<img width="62" alt="Screenshot 2025-03-06 at 10 23 05 AM" src="https://github.com/user-attachments/assets/2171bef0-99b9-4d89-b264-097bbfb3df55" />
<img width="213" alt="Screenshot 2025-03-06 at 10 23 40 AM" src="https://github.com/user-attachments/assets/5334de40-bf53-4368-a387-608a4daf8d57" />

4. Even after I exit the program, 1 + 2 persist.

bcherny reviewed (changes requested):
^

dcolascione commented:
The arrow is the point. It shows the user that a separate region of prompt history has begun. You see the same thing in a shell. You're essentially saying we shouldn't have the feature.

dcolascione commented:
https://github.com/user-attachments/assets/734aa62d-0ac7-40f7-930c-e7d1f8777ba7

Commit:
Correctly strip control sequences in always-prompt region

Commit:
Fix bizarre heights

Commit:
Fix highlighting bug

Commit:
Turn OSC133 support off by default

dcolascione commented:
Should be fixed now --- and off by default. Funny thing is, I didn't even notice the highlighting bug. I thought the weird highlighting you were pointing out was what iTerm2 did, which indeed is the feature.

Commit:
Don't assume height is an integer or even a real number

Commit:
Add missing option check

Commit:
add osc133 option to public interface as well as internal struct

Commit:
Add atomic region marking to reduce flicker

The commit gives supporting terminals (kitty, iTerm2, Wezterm, etc.)
a hint that the ink program will be writing a bunch of output and that
the terminal shouldn't refresh until it's done.  On terminals that
don't support the hint, it does nothing.

Commit:
Send escape codes only to TTYs
